cli/tests/test_touch_command.rs

```diff
@@ -0,0 +1,310 @@
+    │  Bookmarks: c
+    │  Author   : Test User <test.user@example.com> (2001-02-03 08:05:13)
+    │  Committer: Test User <test.user@example.com> (2001-02-03 08:05:20)
+    │
+    │      (no description set)
+    │
+    ○  Commit ID: f121a0fb72e1790e4116b2e3b6989c795ac7f74b
+    │  Change ID: kkmpptxzrspxrzommnulwmwkkqwworpl
+    │  Bookmarks: b
+    │  Author   : Test User <test.user@example.com> (2001-02-03 08:05:20)
+    │  Committer: Test User <test.user@example.com> (2001-02-03 08:05:20)
+    │
+    │      (no description set)
+    │
+    ○  Commit ID: e6086990958c236d72030f0a2651806aa629f5dd
+    │  Change ID: qpvuntsmwlqtpsluzzsnyyzlmlwvmlnu
+    │  Bookmarks: a
+    │  Author   : Test User <test.user@example.com> (2001-02-03 08:05:09)
+    │  Committer: Test User <test.user@example.com> (2001-02-03 08:05:09)
+    │
+    │      (no description set)
+    │
+    ◆  Commit ID: 0000000000000000000000000000000000000000
+       Change ID: zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz
+       Author   : (no name set) <(no email set)> (1970-01-01 11:00:00)
+       Committer: (no name set) <(no email set)> (1970-01-01 11:00:00)
+
+           (no description set)
+
+    [EOF]
+    ");
+
+    // Set author
+    work_dir.run_jj(["undo"]).success();
+    work_dir
+        .run_jj([
+            "touch",
+            "--author",
+            "Alice <alice@example.com>",
+            "kkmpptxzrspx",
+        ])
+        .success();
+    insta::assert_snapshot!(get_log(&work_dir), @r"
+    @  Commit ID: 74007c679b9e4f13d1e3d553ef8397586b033421
+    │  Change ID: mzvwutvlkqwtuzoztpszkqxkqmqyqyxo
+    │  Bookmarks: c
+    │  Author   : Test User <test.user@example.com> (2001-02-03 08:05:13)
+    │  Committer: Test User <test.user@example.com> (2001-02-03 08:05:23)
+    │
+    │      (no description set)
+    │
+    ○  Commit ID: d070c8adbc590813c81e296591d6b2cac8f3bb41
+    │  Change ID: kkmpptxzrspxrzommnulwmwkkqwworpl
+    │  Bookmarks: b
+    │  Author   : Alice <alice@example.com> (2001-02-03 08:05:11)
+    │  Committer: Test User <test.user@example.com> (2001-02-03 08:05:23)
+    │
+    │      (no description set)
+    │
+    ○  Commit ID: e6086990958c236d72030f0a2651806aa629f5dd
+    │  Change ID: qpvuntsmwlqtpsluzzsnyyzlmlwvmlnu
+    │  Bookmarks: a
+    │  Author   : Test User <test.user@example.com> (2001-02-03 08:05:09)
+    │  Committer: Test User <test.user@example.com> (2001-02-03 08:05:09)
+    │
+    │      (no description set)
+    │
+    ◆  Commit ID: 0000000000000000000000000000000000000000
+       Change ID: zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz
+       Author   : (no name set) <(no email set)> (1970-01-01 11:00:00)
+       Committer: (no name set) <(no email set)> (1970-01-01 11:00:00)
+
+           (no description set)
+
+    [EOF]
+    ");
+
+    // New change-id
+    work_dir.run_jj(["undo"]).success();
+    work_dir
+        .run_jj(["touch", "--update-change-id", "kkmpptxzrspx"])
+        .success();
```

Comment:
If we snapshot the output of `jj touch`, we'll get weird result because descendants will be rebased twice.

Reply:
Ho, I struggled with this, which made me move `set_rewritten_commit()` in multiple places. Maybe the code that set the parents wasn't a leftover…
I should have added a test, though.

Reply:
So I think we'll need `commit_builder` flag to not assume generated change id means the commit was duplicated.

```rust
        mut_repo.set_predecessors(commit.id().clone(), predecessors);
        if let Some(rewrite_source) = self.rewrite_source {
            if rewrite_source.change_id() == commit.change_id() { // XXX replace this with some boolean?
                mut_repo.set_rewritten_commit(rewrite_source.id().clone(), commit.id().clone());
            }
        }
```

Alternatively, we can make `commit_builder.write()` release `mut_repo`.

Reply:
I think it's better to merge #7231 first because we can get rid of the hack.